File: src/api/db/table.rs
use std::error::Error;

use super::{table_query_builder::TableQueryBuilder, table_row::TableRow, TableQuery};

pub trait IntoTable<TableType: Table> {
    fn into_table(&self) -> TableType;
}

pub trait Table {
    type IdType;
    fn save_all(&self, data: Vec<TableRow<Self::IdType>>) -> Result<(), Box<dyn Error>>;
    fn find_by_query(
        &self,
        query: TableQuery,
    ) -> Result<Vec<TableRow<Self::IdType>>, Box<dyn Error>>;
    fn find_by<'a>(&'a self) -> TableQueryBuilder<'a, Self::IdType>
    where
        Self: Sized,
    {
        TableQueryBuilder::<'a, Self::IdType>::new(self)
    }
}

File: src/api/db/Table.ts
import { TableQuery } from "./TableQuery";
import { TableQueryBuilder } from "./TableQueryBuilder";
import { TableRow } from "./TableRow";

export type Table<IdType> = {
    saveAll(data: TableRow<IdType>[]): Promise<void>;
    findByQuery(query: TableQuery): Promise<TableRow<IdType>[]>;
    findBy(): TableQueryBuilder<IdType>;
}